src/test/validation_chainstatemanager_tests.cpp

```diff
@@ -104,4 +104,58 @@ BOOST_AUTO_TEST_CASE(chainstatemanager)
     WITH_LOCK(::cs_main, manager.Unload());
 }
 
+//! Test rebalancing the caches associated with each chainstate.
+BOOST_AUTO_TEST_CASE(chainstatemanager_rebalance_caches)
+{
+    ChainstateManager manager;
+    int max_cache = 10000;
+    manager.m_total_coinsdb_cache = max_cache;
+    manager.m_total_coinstip_cache = max_cache;
+
+    std::vector<CChainState*> chainstates;
+
+    // Create a legacy (IBD) chainstate.
+    //
+    ENTER_CRITICAL_SECTION(cs_main);
+    CChainState& c1 = manager.InitializeChainstate();
+    LEAVE_CRITICAL_SECTION(cs_main);
```

Comment:
style-nit in commit 1e6e0b9af6c5470dbf1e61:

Now that you have to force push anyway, might as well remove the clumsy `ENTER_CRITICAL_SECTION`?

```suggestion
    CChainState& c1 = *WITH_LOCK(cs_main, return &manager.InitializeChainstate());
```